Mining interface followups, reduce cs_main locking, test rpc bug fix

Sjors opened the pull request:
Followups from #30200

Fixes:
- `std::unique_ptr` needs `#include <memory>` (noticed while working on #30332, which has fewer includes than its parent PR that I originally tested with)
- Drop lock from createNewBlock that was spuriously added
- Have testBlockValidity hold cs_main instead of caller (also fixes a race condition in test-only code)

Refactor:
- Use CHECK_NONFATAL to avoid single-use symbol (refactor)
- move output argument `state` to the end of `testBlockValidity`, see https://github.com/bitcoin/bitcoin/pull/30200#discussion_r1647987176

Sjors committed:
Add missing include for mining interface

Needed for std::unique_ptr

Sjors committed:
refactor: testBlockValidity make out argument last

Sjors commented on src/interfaces/mining.h:
```diff
@@ -44,6 +45,7 @@ class Mining
      * @returns a block template
      */
     virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool = true) = 0;
+
```
Added blank line here for consistency.

itornaza reviewed (approved):
tested ACK 22fe97a5cc612f438ec6280757487934501bf527

Manually validated all instances of `testBlockValidity` in the entire code base are having their parameter list updated with `state` as their last output parameter using good old `git grep -n testBlockValidity`. Now, the interface can be used with the `libmultiprocess` code generator!

As for the standard testing, I got the pr from upstream, built and run all unit and functional tests including the entire extended test harness and all tests pass.

(Special thanks for incorporating that empty line, it makes eyeballing of different methods pretty easier to spot even for people with poorer eyesight!)

AngusP reviewed (approved):
Tested ACK 22fe97a5cc612f438ec6280757487934501bf527

tdb3 reviewed (approved):
ACK 22fe97a5cc612f438ec6280757487934501bf527

Changes look straightforward.  Took a quick look at the other interface functions in `mining.h`, and the function arguments look to all now be in the in/inout/out order.  Also manually ran `getblocktemplate` RPC on signet and on regtest (after broadcasting a few transactions).  All looked well.

maflcko reviewed (commented):
`CHECK_NONFATAL` can be used to write shorter code and not pollute the scope with single-use symbols. That is:

```cpp
    std::optional<uint256> maybe_tip{miner.getTipHash()};
    CHECK_NONFATAL(maybe_tip);
    uint256 tip{maybe_tip.value()};
```

can be:

```cpp
    uint256 tip{CHECK_NONFATAL(miner.getTipHash()).value()};

AngusP commented on src/rpc/mining.cpp:
```diff
@@ -390,7 +390,7 @@ static RPCHelpMan generateblock()
         LOCK(cs_main);
```
It is also acquired within `miner.testBlockValidity` (and in `miner.createNewBlock`) so maybe is now unnecessary outside the Miner interface?

https://github.com/bitcoin/bitcoin/blob/a9716c53f05082d6d89ebea51a46d4404efb12d7/src/node/interfaces.cpp#L873-L877

Sjors replied:
I pushed 1f0559e9954253c23af7ee9b49e57eb3d7a45bea to drop it from `createNewBlock`. Will look at the other occurances too.

Sjors replied:
The lock at line 390 has been there from the introduction of the `generateblock` RPC call in #17693. I don't see any discussion of it in the reviews.

I think it's because `assert(pindexPrev && pindexPrev == chainstate.m_chain.Tip());` inside of `TestBlockValidity`. Added d23968e0fb6e1d7c5d19b07722796181d6e04f5b to explain that, and also make `testBlockValidity` require the lock.

Sjors replied:
Actually that assert isn't the problem, it merely checks that the `chainman().ActiveChainstate()` and `chainman().ActiveChain().Tip()` that we passed in match.

The problem is deeper down the call stack, where `TestBlockValidity` calls `ConnectBlock`. That takes a `viewNew` argument which is `chainstate.CoinsTip()`, and our proposed `block`, and then it does `assert(hashPrevBlock == view.GetBestBlock());`.

Sjors replied:
Pushed c8343e29a66082b37c37ee0ff8f7433b97eea9b1 to deal with this (replaces https://github.com/bitcoin/bitcoin/commit/18d6b0f86769f89738a3b08aa8152a8479cef5d4). This drops both locks from the RPC code.

An alternative that would actually prevent the tip from updating, rather than fail the RPC call, is to hold the lock even longer. But callers need to handle failure anyway since `testBlockValidity` can fail for other reasons.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30335#discussion_r1654338332

> What is this lock used for?

Marking this thread resolved as the lock is now removed in later commit c8343e29a66082b37c37ee0ff8f7433b97eea9b1

Sjors committed:
Drop unneeded lock from createNewBlock

This was added in 4bf2e361da1964f7c278b4939967a0e5afde20b0, but
BlockAssembler::CreateNewBlock already locks cs_main internally.

Sjors committed:
refactor: use CHECK_NONFATAL to avoid single-use symbol

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -870,9 +870,10 @@ class MinerImpl : public Mining
     }
 
-    bool testBlockValidity(BlockValidationState& state, const CBlock& block, bool check_merkle_root) override
+    bool testBlockValidity(const CBlock& block, bool check_merkle_root, BlockValidationState& state) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) override
```
I initially added it only to the header, but that resulted in:

```
node/interfaces.cpp:875:9: warning: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::recursive_mutex>>' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        AssertLockHeld(cs_main);
        ^
```


Should I add it in both? Or declare the `override` first (which would need a new header).

Sjors replied:
Good point, it doesn't make sense to require the lock. Will try something else.

Sjors committed:
Have testBlockValidity hold cs_main instead of caller

The goal of interfaces is to eventually run in their own process,
so we can't use EXCLUSIVE_LOCKS_REQUIRED in their declaration.

However TestBlockValidaty will crash (in its call to ConnectBlock)
if the tip changes from under the proposed block.

Have the testBlockValidity implementation  hold the lock instead,
and non-fatally check for this condition.

AngusP commented on src/node/interfaces.cpp:
```diff
@@ -872,8 +872,12 @@ class MinerImpl : public Mining
+        LOCK(cs_main);
+        CBlockIndex* tip{chainman().ActiveChain().Tip()};
+        // Fail if the tip updated before the lock was taken
+        if (block.hashPrevBlock != tip->GetBlockHash()) return false;
```
Currently if this happens as you said https://github.com/bitcoin/bitcoin/pull/30335#discussion_r1654637548 it looks like we'd have crashed which is fun

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -881,7 +881,6 @@ class MinerImpl : public Mining
         BlockAssembler::Options options;
         ApplyArgsManOptions(gArgs, options);
 
-        LOCK(::cs_main);
```
In commit "Drop unneeded lock from createNewBlock" (1e4918b8f3af20577acdc9f201af44153e29bcb3)

Commit message could mention the reason the lock is not required in CreateNewBlock, which is just that CreateNewBlock already locks cs_main internally.

Sjors replied:
Reworded

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -870,18 +870,21 @@ class MinerImpl : public Mining
+        LOCK(cs_main);
+        CBlockIndex* tip{chainman().ActiveChain().Tip()};
+        // Fail if the tip updated before the lock was taken
+        if (block.hashPrevBlock != tip->GetBlockHash()) return false;
```
In commit "Have testBlockValidity hold cs_main instead of caller" (c8343e29a66082b37c37ee0ff8f7433b97eea9b1)

re: https://github.com/bitcoin/bitcoin/pull/30335#discussion_r1654687943

> `state` will be valid, no? So this may be confusing?

It seems like it would be better to do something like:


```c++
if (block.hashPrevBlock != tip->GetBlockHash()) {
    state.Error("Block does not connect to current chain tip.")
    return false;
}
```

Sjors replied:
Taken

ryanofsky reviewed (approved):
Code review ACK c8343e29a66082b37c37ee0ff8f7433b97eea9b1

Nice changes getting rid of some over-broad and recursive uses of cs_main.

ryanofsky commented:
It would be good to mention that the are some changes to cs_main locking (including a potential bugfix) in the PR description, otherwise this should be a refactoring with no changes in behavior.

Sjors commented:
Addressed feedback and expanded PR description.

AngusP reviewed (approved):
Code Review ACK a74b0f93efa1d9eaf5abc2f6591c44a632aec6ed

itornaza reviewed (approved):
Tested ACK a74b0f93efa1d9eaf5abc2f6591c44a632aec6ed

Code reviewed and validated all the initial refactor and appended fix changes as they were discovered during the course of this PR.

Cleaned up and rebuilt this commit running the unit, functional and extended tests that all pass.

ryanofsky reviewed (approved):
Code review ACK a74b0f93efa1d9eaf5abc2f6591c44a632aec6ed. Just new error string is added since last review, and a commit message was updated

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -667,9 +663,7 @@ static RPCHelpMan getblocktemplate()
-    std::optional<uint256> maybe_tip{miner.getTipHash()};
-    CHECK_NONFATAL(maybe_tip);
-    uint256 tip{maybe_tip.value()};
+    uint256 tip{CHECK_NONFATAL(miner.getTipHash()).value()};
```
Added to #30356.
